lib/workflow/postgres/postgres.go

```diff
@@ -0,0 +1,726 @@
+
+func NewPostgresBackend(connURI string) *PostgresBackend {
+	return &PostgresBackend{
+		connURI: connURI,
+	}
+}
+
+func (pg *PostgresBackend) Init() error {
+	db, err := sql.Open("postgres", pg.connURI)
+	if err != nil {
+		return err
+	}
+	pg.db = db
+
+	var dbName string
+	err = pg.db.QueryRow("SELECT CURRENT_DATABASE()").Scan(&dbName)
+	if err != nil {
+		return errors.Wrap(err, "could not get database name")
+	}
+
+	dbInstance, err := postgres.WithInstance(pg.db, &postgres.Config{
+		MigrationsTable: "workflow_schema_version",
+	})
+	if err != nil {
+		return err
+	}
+
+	sourceDrv, err := source.Open("array://local")
+	if err != nil {
+		return err
+	}
+
+	m, err := migrate.NewWithInstance("built-in schema", sourceDrv, dbName, dbInstance)
+	if err != nil {
+		return err
+	}
+	// defer m.Close() I don't think we want to call close here because it'll close our db instance
+
+	err = m.Up()
+	if err == migrate.ErrNoChange {
+		logrus.Info("No database migration required")
+	} else if err != nil {
+		return errors.Wrap(err, "migration failed")
+	}
+
+	return nil
+}
+
+func (pg *PostgresBackend) Close() error {
+	if pg.db != nil {
+		return pg.db.Close()
+	}
+	return nil
+}
+
+func (pg *PostgresBackend) GetScheduledWorkflowParameters(ctx context.Context, scheduleName string, workflowName string) ([]byte, error) {
+	row := pg.db.QueryRowContext(ctx, "SELECT parameters FROM recurring_workflow_schedules WHERE name = $1 and workflow_name = $2",
+		scheduleName, workflowName)
+	var data []byte
+
+	err := row.Scan(&data)
+	if err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+func (pg *PostgresBackend) GetScheduledWorkflowRecurrence(ctx context.Context, scheduleName string, workflowName string) (string, error) {
+	row := pg.db.QueryRowContext(ctx, "SELECT recurrence FROM recurring_workflow_schedules WHERE name = $1 and workflow_name = $2",
+		scheduleName, workflowName)
+
+	var data string
+	err := row.Scan(&data)
+	if err != nil {
+		return "", err
+	}
+
+	return data, nil
+}
+
+func (pg *PostgresBackend) ListWorkflowSchedules(ctx context.Context) ([]*backend.Schedule, error) {
+	ctx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
+
+	rows, err := pg.db.QueryContext(ctx, listRecurringWorkflowsQuery)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not query recurring workflows")
+	}
+
+	defer rows.Close() // nolint: errcheck
+
+	schedules := make([]*backend.Schedule, 0)
+	for rows.Next() {
+		var scheduledWorkflow backend.Schedule
+		err := rows.Scan(
+			&scheduledWorkflow.ID,
+			&scheduledWorkflow.Enabled,
+			&scheduledWorkflow.Name,
+			&scheduledWorkflow.WorkflowName,
+			&scheduledWorkflow.Parameters,
+			&scheduledWorkflow.Recurrence,
+			&scheduledWorkflow.NextDueAt,
+			&scheduledWorkflow.LastStart,
+			&scheduledWorkflow.LastEnd,
+		)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not read workflow schedules")
+		}
+		schedules = append(schedules, &scheduledWorkflow)
+	}
```

Comment:
Here we need to check rows.Err() since Next() returns false on the end of the results AND on an error.